### PR TITLE
MM-20755: fix post actions in DMs/GMs

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
The fix for https://github.com/mattermost/mattermost-server/issues/12377 assumed that all channels have teams, but this is false for DMs and GMs.  Test for this and avoid failing on a missing team as such.

#### Ticket Link
Fixes: https://mattermost.atlassian.net/browse/MM-20755

### CODE REVIEW

File: app/integration_action_test.go [hunk 1]
```diff
@@ -68,250 +68,282 @@ func TestPostActionInvalidURL(t *testing.T) {
...
+
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				request := model.PostActionIntegrationRequestFromJson(r.Body)
+				assert.NotNil(t, request)
+
+				assert.Equal(t, request.UserId, th.BasicUser.Id)
+				assert.Equal(t, request.UserName, th.BasicUser.Username)
+				assert.Equal(t, request.ChannelId, channel.Id)
+				assert.Equal(t, request.ChannelName, channel.Name)
+				if channel.Type == model.CHANNEL_DIRECT || channel.Type == model.CHANNEL_GROUP {
```
Comment (lieut-data) [inline app/integration_action_test.go]: The diff here is painful, but the only meaningful change is here, plus parameterizing the channel throughout the tests.

File: app/integration_action.go [hunk 2]
```diff
@@ -171,12 +179,13 @@ func (a *App) DoPostActionWithCookie(postId, actionId, userId, selectedOption st
...
 
-	tr := <-teamChan
-	if tr.Err != nil {
+	tr, ok := <-teamChan
+	if ok && tr.Err != nil {
 		return "", tr.Err
+	} else if ok {
+		team := tr.Data.(*model.Team)
+		upstreamRequest.TeamName = team.Name
 	}
```
Comment (hanzei) [inline app/integration_action.go]: I think we can make the code a bit cleaner, if we drop the double check of `ok` and nest the `if`:
```go
	tr, ok := <-teamChan
	if ok {
		if tr.Err != nil {
			return "", tr.Err
		}

		team := tr.Data.(*model.Team)
		upstreamRequest.TeamName = team.Name
	}
```

File: app/integration_action_test.go [hunk 3]
```diff
@@ -68,250 +68,282 @@ func TestPostActionInvalidURL(t *testing.T) {
...
-			assert.Equal(t, request.DataSource, "")
-		}
-		assert.Equal(t, "foo", request.Context["s"])
-		assert.EqualValues(t, 3, request.Context["n"])
-		fmt.Fprintf(w, `{"post": {"message": "updated"}, "ephemeral_text": "foo"}`)
-	}))
-	defer ts.Close()
+	testCases := []struct {
+		Description string
+		Channel     func(th *TestHelper) *model.Channel
```
Comment (hanzei) [inline app/integration_action_test.go]: Nice solution :+1:

**Comment by hanzei** [comment]
Out of scope of this PR, but I would like to see a e2e test for post actions. I looked into the existing ones and only found a test for interactive menus.

@saturninoabril Is this something that is planed or should we open a ticket for this?

**Comment by saturninoabril** [comment]
@hanzei Thanks for checking. Yes, interactive message buttons  are planned to be written into Cypress (basically all in the release testing spreadsheet). But would be good to open a ticket so we could prioritize work on it (just include specific test case in the ticket description).

**Comment by hanzei** [comment]
@saturninoabril I've filled https://mattermost.atlassian.net/browse/MM-20810, but I'm a little unsure what should go into the ticket. Feel free to edit it.

**Comment by saturninoabril** [comment]
Thanks @hanzei!  I'll take care of that.

**Comment by amyblais** [comment]
@cpoile @prapti Kind reminder to review this for 5.18.

**Comment by lieut-data** [comment]
CircleCI builds are flaky: merging this one based on the Jenkins pass.

**Pull request merged**